In [34]:
pip install imbalanced-learn

     |████████████████████████████████| 167 kB 4.5 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd
url="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_d67855dfe2ec483dbb1f6c2471528b19.csv"
df=pd.read_csv(url,header=None,na_values=["?"])
df.head()

0     1  2  3  4  5  6  7  8  9   ... 16   17 18    19 20    21  \
0  hypothyroid  72.0  M  f  f  f  f  f  f  f  ...  y  0.6  y  15.0  y  1.48   
1  hypothyroid  15.0  F  t  f  f  f  f  f  f  ...  y  1.7  y  19.0  y  1.13   
2  hypothyroid  24.0  M  f  f  f  f  f  f  f  ...  y  0.2  y   4.0  y  1.00   
3  hypothyroid  24.0  F  f  f  f  f  f  f  f  ...  y  0.4  y   6.0  y  1.04   
4  hypothyroid  77.0  M  f  f  f  f  f  f  f  ...  y  1.2  y  57.0  y  1.28   

  22    23 24  25  
0  y  10.0  n NaN  
1  y  17.0  n NaN  
2  y   0.0  n NaN  
3  y   6.0  n NaN  
4  y  44.0  n NaN  

[5 rows x 26 columns]

In [2]:
df.iloc[:,0].groupby(df.iloc[:,0]).count()

0
hypothyroid     151
negative       3012
Name: 0, dtype: int64

In [3]:
dtypes=pd.DataFrame(df.dtypes).rename(columns={0:"Type"})
dtypes

Type
0    object
1   float64
2    object
3    object
4    object
5    object
6    object
7    object
8    object
9    object
10   object
11   object
12   object
13   object
14   object
15  float64
16   object
17  float64
18   object
19  float64
20   object
21  float64
22   object
23  float64
24   object
25  float64

In [4]:
objects=dtypes[dtypes["Type"]>="object"]
objects

Type
0   object
2   object
3   object
4   object
5   object
6   object
7   object
8   object
9   object
10  object
11  object
12  object
13  object
14  object
16  object
18  object
20  object
22  object
24  object

In [5]:
len(objects)

19

In [6]:
objects=df.iloc[:,objects.index.tolist()]

In [7]:
objects.head()

0  2  3  4  5  6  7  8  9  10 11 12 13 14 16 18 20 22 24
0  hypothyroid  M  f  f  f  f  f  f  f  f  f  f  f  y  y  y  y  y  n
1  hypothyroid  F  t  f  f  f  f  f  f  f  f  f  f  y  y  y  y  y  n
2  hypothyroid  M  f  f  f  f  f  f  f  f  f  f  f  y  y  y  y  y  n
3  hypothyroid  F  f  f  f  f  f  f  f  f  f  f  f  y  y  y  y  y  n
4  hypothyroid  M  f  f  f  f  f  f  f  f  f  f  f  y  y  y  y  y  n

In [8]:
objects_missing=pd.DataFrame(objects.isna().sum()/len(objects)).rename(columns={0:"Percentage"})
objects_missing

Percentage
0     0.000000
2     0.023079
3     0.000000
4     0.000000
5     0.000000
6     0.000000
7     0.000000
8     0.000000
9     0.000000
10    0.000000
11    0.000000
12    0.000000
13    0.000000
14    0.000000
16    0.000000
18    0.000000
20    0.000000
22    0.000000
24    0.000000

In [9]:
objects_missing[objects_missing["Percentage"]>0.1]

Empty DataFrame
Columns: [Percentage]
Index: []

In [10]:
objects=objects.fillna(objects.mode())
objects.isna().any()

0     False
2      True
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
16    False
18    False
20    False
22    False
24    False
dtype: bool

In [11]:
objects.iloc[:,1].mode()

0    F
dtype: object

In [12]:
objects=objects.fillna("F")
objects.isna().any()

0     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
16    False
18    False
20    False
22    False
24    False
dtype: bool

In [13]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
objects=objects.apply(le.fit_transform)
objects.head()

0   2   3   4   5   6   7   8   9   10  11  12  13  14  16  18  20  22  24
0   0   1   0   0   0   0   0   0   0   0   0   0   0   1   1   1   1   1   0
1   0   0   1   0   0   0   0   0   0   0   0   0   0   1   1   1   1   1   0
2   0   1   0   0   0   0   0   0   0   0   0   0   0   1   1   1   1   1   0
3   0   0   0   0   0   0   0   0   0   0   0   0   0   1   1   1   1   1   0
4   0   1   0   0   0   0   0   0   0   0   0   0   0   1   1   1   1   1   0

In [14]:
numeric=dtypes[dtypes["Type"]!="object"]
numeric

Type
1   float64
15  float64
17  float64
19  float64
21  float64
23  float64
25  float64

In [15]:
numeric=df.iloc[:,numeric.index.tolist()]
numeric.head()

1      15   17    19    21    23  25
0  72.0   30.0  0.6  15.0  1.48  10.0 NaN
1  15.0  145.0  1.7  19.0  1.13  17.0 NaN
2  24.0    0.0  0.2   4.0  1.00   0.0 NaN
3  24.0  430.0  0.4   6.0  1.04   6.0 NaN
4  77.0    7.3  1.2  57.0  1.28  44.0 NaN

In [16]:
numeric_missing=pd.DataFrame(numeric.isna().sum()/len(numeric)).rename(columns={0:"Percentage"})
numeric_missing

Percentage
1     0.141005
15    0.147961
17    0.219728
19    0.078723
21    0.078407
23    0.078090
25    0.917800

In [17]:
numeric_missing=numeric_missing[numeric_missing["Percentage"]<=0.1]
numeric_missing

Percentage
19    0.078723
21    0.078407
23    0.078090

In [18]:
numeric_missing.index.tolist()

[19, 21, 23]

In [19]:
numeric=df.iloc[:,numeric_missing.index.tolist()]
numeric.head()

19    21    23
0  15.0  1.48  10.0
1  19.0  1.13  17.0
2   4.0  1.00   0.0
3   6.0  1.04   6.0
4  57.0  1.28  44.0

In [20]:
numeric.isna().any()

19    True
21    True
23    True
dtype: bool

In [21]:
numeric=numeric.fillna(numeric.median())
numeric.isna().any()

19    False
21    False
23    False
dtype: bool

In [22]:
df=pd.concat([objects,numeric],axis=1)
df.head()

0   2   3   4   5   6   7   8   9   10  ...  13  14  16  18  20  22  24  \
0   0   1   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
1   0   0   1   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
2   0   1   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
3   0   0   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
4   0   1   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   

     19    21    23  
0  15.0  1.48  10.0  
1  19.0  1.13  17.0  
2   4.0  1.00   0.0  
3   6.0  1.04   6.0  
4  57.0  1.28  44.0  

[5 rows x 22 columns]

In [23]:
df=df.rename(columns={0:"Class"})
df.head()

Class  2  3  4  5  6  7  8  9  10  ...  13  14  16  18  20  22  24    19  \
0      0  1  0  0  0  0  0  0  0   0  ...   0   1   1   1   1   1   0  15.0   
1      0  0  1  0  0  0  0  0  0   0  ...   0   1   1   1   1   1   0  19.0   
2      0  1  0  0  0  0  0  0  0   0  ...   0   1   1   1   1   1   0   4.0   
3      0  0  0  0  0  0  0  0  0   0  ...   0   1   1   1   1   1   0   6.0   
4      0  1  0  0  0  0  0  0  0   0  ...   0   1   1   1   1   1   0  57.0   

     21    23  
0  1.48  10.0  
1  1.13  17.0  
2  1.00   0.0  
3  1.04   6.0  
4  1.28  44.0  

[5 rows x 22 columns]

In [24]:
df.groupby("Class")["Class"].count()

Class
0     151
1    3012
Name: Class, dtype: int64

0 = Hypothyroid; 1 = Negative

In [25]:
X=df.drop(["Class"],axis=1)
X.head()

2   3   4   5   6   7   8   9   10  11  ...  13  14  16  18  20  22  24  \
0   1   0   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
1   0   1   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
2   1   0   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
3   0   0   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   
4   1   0   0   0   0   0   0   0   0   0  ...   0   1   1   1   1   1   0   

     19    21    23  
0  15.0  1.48  10.0  
1  19.0  1.13  17.0  
2   4.0  1.00   0.0  
3   6.0  1.04   6.0  
4  57.0  1.28  44.0  

[5 rows x 21 columns]

In [26]:
y=df[["Class"]]
y.head()

Class
0      0
1      0
2      0
3      0
4      0

In [27]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X=pd.DataFrame(scaler.fit_transform(X))
X.head()

0         1         2         3         4         5         6   \
0  1.575906 -0.413055 -0.133027 -0.116005 -0.184386 -0.287189 -0.288477   
1 -0.634556  2.420986 -0.133027 -0.116005 -0.184386 -0.287189 -0.288477   
2  1.575906 -0.413055 -0.133027 -0.116005 -0.184386 -0.287189 -0.288477   
3 -0.634556 -0.413055 -0.133027 -0.116005 -0.184386 -0.287189 -0.288477   
4  1.575906 -0.413055 -0.133027 -0.116005 -0.184386 -0.287189 -0.288477   

         7         8         9   ...        11        12        13        14  \
0 -0.142557 -0.179752 -0.113173  ... -0.179752  0.416719  0.530664  0.292318   
1 -0.142557 -0.179752 -0.113173  ... -0.179752  0.416719  0.530664  0.292318   
2 -0.142557 -0.179752 -0.113173  ... -0.179752  0.416719  0.530664  0.292318   
3 -0.142557 -0.179752 -0.113173  ... -0.179752  0.416719  0.530664  0.292318   
4 -0.142557 -0.179752 -0.113173  ... -0.179752  0.416719  0.530664  0.292318   

        15        16       17        18        19        20  
0  0.29168  0.291041 -0.29927 -2.140310  2.313337 -1.809839  
1  0.29168  0.291041 -0.29927 -2.048715  0.704388 -1.688885  
2  0.29168  0.291041 -0.29927 -2.392197  0.106779 -1.982629  
3  0.29168  0.291041 -0.29927 -2.346399  0.290659 -1.878955  
4  0.29168  0.291041 -0.29927 -1.178560  1.393938 -1.222352  

[5 rows x 21 columns]

In [28]:
import numpy as np
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

# feature extraction. We are going to output the selection scores for all features and select the features with the highest scores.
test = SelectKBest(score_func=f_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
# summarize scores
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

scores
18  515.501135
20  344.486595
12   25.311835
19   21.713726
13   14.981797
14   13.599430
15   13.539914
16   13.480440
17    8.182420
5     7.141318
4     5.550795
1     3.583723
2     2.806814
6     2.075553
9     2.031016
8     1.704588
7     1.435605
0     0.971374
3     0.535925
11    0.371975
10    0.100269

In [29]:
X_scores=scores[scores["scores"]>=1]
len(X_scores)

17

In [30]:
X=X.iloc[:,X_scores.index]
X.head()

18        20        12        19        13        14       15  \
0 -2.140310 -1.809839  0.416719  2.313337  0.530664  0.292318  0.29168   
1 -2.048715 -1.688885  0.416719  0.704388  0.530664  0.292318  0.29168   
2 -2.392197 -1.982629  0.416719  0.106779  0.530664  0.292318  0.29168   
3 -2.346399 -1.878955  0.416719  0.290659  0.530664  0.292318  0.29168   
4 -1.178560 -1.222352  0.416719  1.393938  0.530664  0.292318  0.29168   

         16       17        5         4         1         2         6   \
0  0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
1  0.291041 -0.29927 -0.287189 -0.184386  2.420986 -0.133027 -0.288477   
2  0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
3  0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
4  0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   

         9         8         7   
0 -0.113173 -0.179752 -0.142557  
1 -0.113173 -0.179752 -0.142557  
2 -0.113173 -0.179752 -0.142557  
3 -0.113173 -0.179752 -0.142557  
4 -0.113173 -0.179752 -0.142557

In [31]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.33,random_state=42)
print(X_train.shape)
print(X_test.shape)

(2119, 17)
(1044, 17)


In [32]:
y_train.groupby("Class")["Class"].count()

Class
0     105
1    2014
Name: Class, dtype: int64

In [33]:
y_test.groupby("Class")["Class"].count()

Class
0     46
1    998
Name: Class, dtype: int64

In [34]:
print("Percentage of majority class in train data: ",2014/2119)
print("Percentage of majority class in test data: ",998/1044)

Percentage of majority class in train data:  0.9504483246814536
Percentage of majority class in test data:  0.9559386973180076


In [35]:
from imblearn.ensemble import BalancedRandomForestClassifier
import numpy as np
from sklearn import model_selection
from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

In [36]:
BForest=BalancedRandomForestClassifier(n_estimators=200,criterion="entropy",max_depth=5,min_samples_split=4,
                                       min_samples_leaf=2,max_features="log2",bootstrap=False,
                                       sampling_strategy="not majority",replacement=True,n_jobs=-1,warm_start=False,
                                       random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
BForest

BalancedRandomForestClassifier(bootstrap=False, criterion='entropy',
                               max_depth=5, max_features='log2',
                               min_samples_leaf=2, min_samples_split=4,
                               n_estimators=200, n_jobs=-1, random_state=0,
                               replacement=True,
                               sampling_strategy='not majority')

In [37]:
from sklearn.metrics import accuracy_score
train_pred=BForest.predict(X_train)
test_pred=BForest.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9863142991977348
Accuracy on test data:  0.9770114942528736


Majority class in both train and test data is about 95%. Hence, for the model to have any significance, accuracy must be above 95% on both train and test data. This is a dataset with highly imbalanced class distribution.

Let's try another classification algorithm and see if we could get past 98.63% accuracy on train data and 97.7% accuracy on test data.

In [38]:
from imblearn.ensemble import BalancedBaggingClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn import svm
from sklearn.naive_bayes import GaussianNB
BBagging=BalancedBaggingClassifier(base_estimator=GradientBoostingClassifier(n_estimators=150),
                                    n_estimators=30,bootstrap_features=True,sampling_strategy="not majority",
                                    n_jobs=-1,random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
BBagging

BalancedBaggingClassifier(base_estimator=GradientBoostingClassifier(n_estimators=150),
                          bootstrap_features=True, n_estimators=30, n_jobs=-1,
                          random_state=0, sampling_strategy='not majority')

In [39]:
train_pred=BBagging.predict(X_train)
test_pred=BBagging.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9957527135441245
Accuracy on test data:  0.9808429118773946


In [40]:
BBagging_svm=BalancedBaggingClassifier(base_estimator=svm.SVC(kernel="linear"),
                                    n_estimators=30,bootstrap_features=True,sampling_strategy="not majority",
                                    n_jobs=-1,random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
BBagging_svm

BalancedBaggingClassifier(base_estimator=SVC(kernel='linear'),
                          bootstrap_features=True, n_estimators=30, n_jobs=-1,
                          random_state=0, sampling_strategy='not majority')

In [41]:
test_pred=BBagging_svm.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9957527135441245
Accuracy on test data:  0.9789272030651341


Of the three classification algorithms above, the BalancedBaggingClassifer with GradientBoostingClassifier as base estimator is the most promising one.

In [42]:
BBagging=BalancedBaggingClassifier(base_estimator=GradientBoostingClassifier(n_estimators=150),
                                    n_estimators=30,bootstrap_features=True,sampling_strategy="not majority",
                                    n_jobs=-1,random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
BBagging

BalancedBaggingClassifier(base_estimator=GradientBoostingClassifier(n_estimators=150),
                          bootstrap_features=True, n_estimators=30, n_jobs=-1,
                          random_state=0, sampling_strategy='not majority')

In [43]:
train_pred=BBagging.predict(X_train)
test_pred=BBagging.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9957527135441245
Accuracy on test data:  0.9808429118773946


In [44]:
from sklearn.metrics import roc_auc_score
print("AUC on train data: ",roc_auc_score(y_train,train_pred))
print("AUC on test data: ",roc_auc_score(y_test,test_pred))

AUC on train data:  0.9571428571428571
AUC on test data:  0.8655572013592402


In [45]:
from sklearn.metrics import confusion_matrix
print("Confusion on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion on test data: \n",confusion_matrix(y_test,test_pred))

Confusion on train data: 
 [[  96    9]
 [   0 2014]]
Confusion on test data: 
 [[ 34  12]
 [  8 990]]


Let's find out which instances from the train and test data that the model missed during the initial round of training and testing.

In [46]:
train_pred=pd.DataFrame(train_pred).rename(columns={0:"Predicted train label"})

In [47]:
y_train=y_train.reset_index(drop=True)

In [48]:
compare_train=pd.concat([y_train,train_pred],axis=1)
compare=pd.DataFrame(compare_train["Class"]==compare_train["Predicted train label"]).rename(columns={0:"Comparison"})
train_comparison=compare[compare["Comparison"]==False]
train_comparison

Comparison
159        False
396        False
572        False
698        False
743        False
1177       False
1242       False
1310       False
1415       False

In [49]:
len(train_comparison)

9

In [50]:
train_indices=train_comparison.index.tolist()
train_indices

[159, 396, 572, 698, 743, 1177, 1242, 1310, 1415]

In [51]:
X_train=X_train.reset_index(drop=True)

In [52]:
train=pd.concat([X_train,y_train],axis=1)
train1=train.loc[train_indices,:]

In [53]:
train1

18        20        12        19        13        14       15  \
159  -1.064066 -0.842214  0.416719 -0.169041  0.530664  0.292318  0.29168   
396  -1.155661 -1.015004  0.416719  0.244689  0.530664  0.292318  0.29168   
572  -1.132763 -0.876772  0.416719 -0.260981  0.530664  0.292318  0.29168   
698  -0.857977 -1.101399  0.416719  1.945577  0.530664  0.292318  0.29168   
743  -0.628989 -0.583028  0.416719  0.152749  0.530664  0.292318  0.29168   
1177 -1.224358 -0.945888  0.416719 -0.215011 -1.884431  0.292318  0.29168   
1242 -0.766382 -0.807656  0.416719  0.612448  0.530664  0.292318  0.29168   
1310 -1.476245 -1.135957 -2.399697 -0.352921 -1.884431  0.292318  0.29168   
1415 -1.201459 -0.876772  0.416719 -0.490831  0.530664  0.292318  0.29168   

            16       17         5         4         1         2         6  \
159   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
396   0.291041 -0.29927 -0.287189  5.423418 -0.413055 -0.133027 -0.288477   
572   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
698   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
743   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
1177  0.291041 -0.29927 -0.287189 -0.184386  2.420986 -0.133027 -0.288477   
1242  0.291041 -0.29927 -0.287189  5.423418 -0.413055 -0.133027 -0.288477   
1310  0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
1415  0.291041 -0.29927  3.482023 -0.184386 -0.413055 -0.133027 -0.288477   

             9         8         7  Class  
159  -0.113173 -0.179752 -0.142557      0  
396  -0.113173 -0.179752 -0.142557      0  
572  -0.113173 -0.179752 -0.142557      0  
698  -0.113173 -0.179752 -0.142557      0  
743  -0.113173 -0.179752 -0.142557      0  
1177 -0.113173 -0.179752 -0.142557      0  
1242 -0.113173 -0.179752 -0.142557      0  
1310 -0.113173 -0.179752 -0.142557      0  
1415 -0.113173 -0.179752 -0.142557      0

In [54]:
train1.shape

(9, 18)

In [55]:
test_pred=pd.DataFrame(test_pred).rename(columns={0:"Predicted test label"})

In [56]:
y_test=y_test.reset_index(drop=True)

In [57]:
compare_test=pd.concat([y_test,test_pred],axis=1)
compare=pd.DataFrame(compare_test["Class"]==compare_test["Predicted test label"]).rename(columns={0:"Comparison"})
test_comparison=compare[compare["Comparison"]==False]
test_comparison

Comparison
30         False
84         False
85         False
109        False
170        False
199        False
203        False
208        False
365        False
377        False
475        False
505        False
606        False
750        False
786        False
804        False
832        False
969        False
1031       False
1038       False

In [58]:
test_indices=test_comparison.index.tolist()
test_indices

[30,
 84,
 85,
 109,
 170,
 199,
 203,
 208,
 365,
 377,
 475,
 505,
 606,
 750,
 786,
 804,
 832,
 969,
 1031,
 1038]

In [59]:
X_test=X_test.reset_index(drop=True)

In [60]:
y_test=y_test.reset_index(drop=True)

In [61]:
test=pd.concat([X_test,y_test],axis=1)
test1=test.loc[test_indices,:]

In [62]:
test1

18        20        12        19        13        14       15  \
30   -1.041167 -1.049562  0.416719  0.888268  0.530664  0.292318  0.29168   
84   -0.697685 -0.600307  0.416719  0.014839  0.530664  0.292318  0.29168   
85   -0.651888 -1.274189  0.416719  3.370646  0.530664  0.292318  0.29168   
109  -1.338852 -0.928609  0.416719 -0.720681  0.530664  0.292318  0.29168   
170  -0.857977 -1.101399  0.416719  1.945577  0.530664  0.292318  0.29168   
199  -1.224358 -0.945888  0.416719 -0.215011 -1.884431  0.292318  0.29168   
203  -0.880876 -0.876772  0.416719  0.566478 -1.884431  0.292318  0.29168   
208   0.012178 -0.928609  0.416719  3.738406  0.530664  0.292318  0.29168   
365  -2.254804 -1.827118  0.416719  0.980208  0.530664  0.292318  0.29168   
377  -0.789280 -0.634865  0.416719 -0.123071  0.530664  0.292318  0.29168   
475  -0.995370 -0.807656  0.416719 -0.077101  0.530664  0.292318  0.29168   
505  -2.231905 -1.809839 -2.399697  0.658418 -1.884431  0.292318  0.29168   
606  -0.560292 -0.945888  0.416719  1.899607  0.530664  0.292318  0.29168   
750  -1.041167 -0.997725  0.416719  0.474539  0.530664  0.292318  0.29168   
786  -0.239709 -0.669423  0.416719  1.439908  0.530664  0.292318  0.29168   
804  -1.980018 -1.654327  0.416719  0.658418  0.530664  0.292318  0.29168   
832  -1.270155 -0.963167  0.416719 -0.398891  0.530664  0.292318  0.29168   
969  -0.903775 -0.928609  0.416719  0.750358  0.530664  0.292318  0.29168   
1031 -1.659435 -1.429700  0.416719  0.612448  0.530664  0.292318  0.29168   
1038  2.782933  0.315481  0.416719  3.462586  0.530664  0.292318  0.29168   

            16       17         5         4         1         2         6  \
30    0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
84    0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
85    0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
109   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
170   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
199   0.291041 -0.29927 -0.287189 -0.184386  2.420986 -0.133027 -0.288477   
203   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
208   0.291041 -0.29927 -0.287189 -0.184386  2.420986 -0.133027 -0.288477   
365   0.291041 -0.29927  3.482023 -0.184386 -0.413055 -0.133027 -0.288477   
377   0.291041 -0.29927 -0.287189  5.423418 -0.413055 -0.133027 -0.288477   
475   0.291041 -0.29927  3.482023 -0.184386 -0.413055 -0.133027 -0.288477   
505   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
606   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
750   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
786   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
804   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
832   0.291041 -0.29927  3.482023 -0.184386  2.420986 -0.133027 -0.288477   
969   0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
1031  0.291041 -0.29927 -0.287189 -0.184386 -0.413055 -0.133027 -0.288477   
1038  0.291041 -0.29927  3.482023 -0.184386  2.420986 -0.133027 -0.288477   

             9         8         7  Class  
30   -0.113173 -0.179752 -0.142557      0  
84   -0.113173 -0.179752 -0.142557      0  
85   -0.113173 -0.179752 -0.142557      0  
109  -0.113173 -0.179752 -0.142557      1  
170  -0.113173 -0.179752 -0.142557      0  
199  -0.113173 -0.179752 -0.142557      0  
203  -0.113173 -0.179752 -0.142557      0  
208  -0.113173 -0.179752 -0.142557      0  
365  -0.113173 -0.179752 -0.142557      1  
377  -0.113173 -0.179752 -0.142557      0  
475  -0.113173 -0.179752 -0.142557      0  
505  -0.113173 -0.179752 -0.142557      1  
606  -0.113173  5.563227 -0.142557      0  
750  -0.113173 -0.179752 -0.142557      0  
786  -0.113173 -0.179752 -0.142557      0  
804  -0.113173 -0.179752 -0.142557      1  
832  -0.113173 -0.179752 -0.

In [63]:
test1.shape

(20, 18)

Since there is only 1 class in the missed instances from round 1's train data. I cannot just re-train the model on the missed instances from the inital train data and test it on the missed instances from the initial test data. There must be at least 2 classes. Hence, the missed instances from the initial round's train and test data must be concatenated and split again. 

In [65]:
round2=pd.concat([train1,test1],axis=0)
round2.groupby("Class")["Class"].count()

Class
0    21
1     8
Name: Class, dtype: int64

In [66]:
X1=round2.drop(["Class"],axis=1)
y1=round2[["Class"]]

In [78]:
X_train1,X_test1,y_train1,y_test1=train_test_split(X1,y1,test_size=0.62,random_state=42)
print(X_train1.shape)
print(X_test1.shape)

(11, 17)
(18, 17)


In [79]:
BBagging1=BalancedBaggingClassifier(base_estimator=GradientBoostingClassifier(n_estimators=200,max_features="auto"),
                                    n_estimators=35,bootstrap_features=False,sampling_strategy="all",
                                    n_jobs=-1,random_state=0).fit(X_train1,np.asarray(y_train1).reshape(y_train1.shape[0],))
BBagging1

BalancedBaggingClassifier(base_estimator=GradientBoostingClassifier(max_features='auto',
                                                                    n_estimators=200),
                          n_estimators=35, n_jobs=-1, random_state=0,
                          sampling_strategy='all')

In [80]:
train_pred1=BBagging1.predict(X_train1)
test_pred1=BBagging1.predict(X_test1)
print("Accuracy on train data for the 2nd round: ",accuracy_score(y_train1,train_pred1))
print("Accuracy on test data for the 2nd round: ",accuracy_score(y_test1,test_pred1))

Accuracy on train data for the 2nd round:  1.0
Accuracy on test data for the 2nd round:  0.8333333333333334


In [81]:
print("Confusion on train data for the 2nd round: \n",confusion_matrix(y_train1,train_pred1))
print("Confusion on test data for the 2nd round: \n",confusion_matrix(y_test1,test_pred1))

Confusion on train data for the 2nd round: 
 [[8 0]
 [0 3]]
Confusion on test data for the 2nd round: 
 [[13  0]
 [ 3  2]]


In [82]:
y_test1.groupby("Class")["Class"].count()

Class
0    13
1     5
Name: Class, dtype: int64

1. After the 2nd round, the model deliver perfect accuracy on train data.
2. For test data, during the initial round, there were 1044 instances in test data and the model correctly classified 1024. During the second round, the model correctly classified 14 more instances of from the initial test data. Hence, **final accuracy on test data is 99.52% [ = ((1024+15)/1044)*100]**